codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java

```diff
@@ -0,0 +1,344 @@
+        return connection().local().numActiveStreams() < maxConcurrentStreams;
+    }
+
+    private boolean isExistingStream(int streamId) {
+        return streamId <= connection().local().lastStreamCreated();
+    }
+
+    private static class PendingStream {
+        final ChannelHandlerContext ctx;
+        final int streamId;
+        final Queue<Frame> frames = new ArrayDeque<Frame>(2);
+
+        PendingStream(ChannelHandlerContext ctx, int streamId) {
+            this.ctx = ctx;
+            this.streamId = streamId;
+        }
+
+        void sendFrames() {
+            for (Frame frame : frames) {
+                frame.send(ctx, streamId);
+            }
+        }
+
+        void close(Throwable t) {
+            for (Frame frame : frames) {
+                frame.release(t);
+            }
+        }
+    }
+
+    private abstract static class Frame {
+        final ChannelPromise promise;
+
+        Frame(ChannelPromise promise) {
+            this.promise = promise;
+        }
+
+        /**
+         * Release any resources (features, buffers, ...) associated with the frame.
+         */
+        void release(Throwable t) {
+            if (t == null) {
+                promise.setSuccess();
+            } else {
+                promise.setFailure(t);
+            }
+        }
+
+        abstract void send(ChannelHandlerContext ctx, int streamId);
+    }
+
+    private class HeadersFrame extends Frame {
+        final Http2Headers headers;
+        final int streamDependency;
+        final short weight;
+        final boolean exclusive;
+        final int padding;
+        final boolean endOfStream;
+
+        HeadersFrame(Http2Headers headers, int streamDependency, short weight, boolean exclusive,
+                     int padding, boolean endOfStream, ChannelPromise promise) {
+            super(promise);
+            this.headers = headers;
+            this.streamDependency = streamDependency;
+            this.weight = weight;
+            this.exclusive = exclusive;
+            this.padding = padding;
+            this.endOfStream = endOfStream;
+        }
+
+        @Override
+        void send(ChannelHandlerContext ctx, int streamId) {
+            writeHeaders(ctx, streamId, headers, streamDependency, weight, exclusive, padding,
+                    endOfStream, promise);
+        }
+    }
+
+    private class DataFrame extends Frame {
+        final ByteBuf data;
+        final int padding;
+        final boolean endOfStream;
+
+        DataFrame(ByteBuf data, int padding, boolean endOfStream, ChannelPromise promise) {
+            super(promise);
+            this.data = data;
+            this.padding = padding;
+            this.endOfStream = endOfStream;
+        }
+
+        @Override
+        public void release(Throwable t) {
+            super.release(t);
+            ReferenceCountUtil.safeRelease(data);
+        }
+
+        @Override
+        void send(ChannelHandlerContext ctx, int streamId) {
```

Comment:
why bother? Seems like adding qualifiers is just adding cruft to the method definitions.